File: ainame-j/src/main/java/com/zhiliao/ainame/controller/OrderController.java
package com.zhiliao.ainame.controller;

import com.zhiliao.ainame.dto.ArtDtos.ArtQueryOutDto;
import com.zhiliao.ainame.dto.OrderDtos.OrderInDto;
import com.zhiliao.ainame.dto.OrderDtos.OrderProductRowOutDto;
import com.zhiliao.ainame.dto.OrderDtos.OrderQueryApiResponseDto;
import com.zhiliao.ainame.dto.OrderDtos.OrderQueryRowOutDto;
import com.zhiliao.ainame.entity.OrderHeader;
import com.zhiliao.ainame.entity.OrderLine;
import com.zhiliao.ainame.repository.CompanyRepository;
import com.zhiliao.ainame.repository.OrderHeaderRepository;
import com.zhiliao.ainame.repository.OrderLineRepository;
import com.zhiliao.ainame.repository.ProductRepository;
import jakarta.transaction.Transactional;
import jakarta.validation.Valid;
import java.util.ArrayList;
import java.util.Map;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

@RestController
@RequestMapping("/order")
public class OrderController {
    private final OrderHeaderRepository orders;
    private final OrderLineRepository orderLines;
    private final CompanyRepository companies;
    private final ProductRepository products;

    public OrderController(
        OrderHeaderRepository orders,
        OrderLineRepository orderLines,
        CompanyRepository companies,
        ProductRepository products
    ) {
        this.orders = orders;
        this.orderLines = orderLines;
        this.companies = companies;
        this.products = products;
    }

    @Transactional
    @PostMapping("/create")
    public ArtQueryOutDto create(@Valid @RequestBody OrderInDto data) {
        var order = new OrderHeader();
        order.setOrderNumber(data.orderNumber());
        order.setCompanyId(data.companyId());
        orders.save(order);

        for (var item : data.productList()) {
            var line = new OrderLine();
            line.setOrderId(order.getId());
            line.setProductId(item.id());
            line.setNumber(item.number());
            orderLines.save(line);
        }
        return new ArtQueryOutDto("created successfully");
    }

    @Transactional
    @PutMapping("/update")
    public ResponseEntity<?> update(@Valid @RequestBody OrderInDto data) {
        if (data.id() == null) {
            return ResponseEntity.badRequest().body(Map.of("detail", "缺少订单 id"));
        }
        var order = orders.findById(data.id()).orElse(null);
        if (order == null) {
            return ResponseEntity.badRequest().body(Map.of("detail", "订单不存在"));
        }
        order.setOrderNumber(data.orderNumber());
        order.setCompanyId(data.companyId());
        orders.save(order);

        orderLines.deleteByOrderId(data.id());
        for (var item : data.productList()) {
            var line = new OrderLine();
            line.setOrderId(data.id());
            line.setProductId(item.id());
            line.setNumber(item.number());
            orderLines.save(line);
        }
        return ResponseEntity.ok(new ArtQueryOutDto("updated successfully"));
    }

    @GetMapping("/query")
    public OrderQueryApiResponseDto query() {
        var result = new ArrayList<OrderQueryRowOutDto>();
        for (var order : orders.findAllByOrderByIdAsc()) {
            String companyName = companies.findById(order.getCompanyId()).map(c -> c.getName()).orElse(null);
            var productList = new ArrayList<OrderProductRowOutDto>();
            float total = 0;
            for (var line : orderLines.findByOrderId(order.getId())) {
                var product = products.findById(line.getProductId()).orElse(null);
                float price = product == null || product.getPrice() == null ? 0 : product.getPrice();
                float number = line.getNumber() == null ? 0 : line.getNumber();
                total += price * number;
                productList.add(new OrderProductRowOutDto(
                    line.getProductId(),
                    product == null ? null : product.getName(),
                    line.getNumber(),
                    price
                ));
            }
            result.add(new OrderQueryRowOutDto(
                order.getId(),
                order.getOrderNumber(),
                order.getCompanyId(),
                companyName,
                productList,
                total
            ));
        }
        return new OrderQueryApiResponseDto(result);
    }

    @Transactional
    @DeleteMapping("/delete")
    public ArtQueryOutDto delete(@RequestParam("id") Integer id) {
        orderLines.deleteByOrderId(id);
        orders.deleteById(id);
        return new ArtQueryOutDto("deleted successfully");
    }
}

File: .net-ainame/Zhiliao.Ainame.Api/Controllers/OrderController.cs
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;
using Zhiliao.Ainame.Api.Contracts;
using Zhiliao.Ainame.Api.Data;
using Zhiliao.Ainame.Api.Entities;

namespace Zhiliao.Ainame.Api.Controllers;

/// <summary>
/// 订单 <c>/order/*</c>：先写订单头拿自增 Id，再插多行明细。
/// </summary>
[ApiController]
[Route("order")]
public class OrderController(AppDbContext db) : ControllerBase
{
    /// <summary><c>POST /order/create</c>：第一次 SaveChanges 得到 <see cref="OrderHeader.Id"/>，再插 <see cref="OrderLine"/>。</summary>
    [HttpPost("create")]
    public async Task<ActionResult<ArtQueryOutDto>> Create([FromBody] OrderInDto data, CancellationToken ct)
    {
        try
        {
            var order = new OrderHeader
            {
                OrderNumber = data.OrderNumber,
                CompanyId = data.CompanyId,
                CreatedTime = DateTime.Now
            };
            db.Orders.Add(order);
            await db.SaveChangesAsync(ct).ConfigureAwait(false);

            foreach (var p in data.ProductList)
            {
                db.OrderLines.Add(new OrderLine
                {
                    OrderId = order.Id,
                    ProductId = p.Id,
                    Number = p.Number,
                    CreatedTime = DateTime.Now
                });
            }

            await db.SaveChangesAsync(ct).ConfigureAwait(false);
            return Ok(new ArtQueryOutDto { Data = "created successfully" });
        }
        catch (Exception ex)
        {
            return StatusCode(500, new { detail = ex.Message });
        }
    }

    /// <summary><c>PUT /order/update</c>：body 须含订单 <c>id</c>；先更新头再删旧明细并重插。</summary>
    [HttpPut("update")]
    public async Task<ActionResult<ArtQueryOutDto>> Update([FromBody] OrderInDto data, CancellationToken ct)
    {
        if (data.Id is null)
            return BadRequest(new { detail = "缺少订单 id" });
        try
        {
            await db.Orders.Where(o => o.Id == data.Id)
                .ExecuteUpdateAsync(s => s
                    .SetProperty(o => o.OrderNumber, data.OrderNumber)
                    .SetProperty(o => o.CompanyId, data.CompanyId), ct)
                .ConfigureAwait(false);

            await db.OrderLines.Where(l => l.OrderId == data.Id).ExecuteDeleteAsync(ct).ConfigureAwait(false);

            foreach (var p in data.ProductList)
            {
                db.OrderLines.Add(new OrderLine
                {
                    OrderId = data.Id.Value,
                    ProductId = p.Id,
                    Number = p.Number,
                    CreatedTime = DateTime.Now
                });
            }

            await db.SaveChangesAsync(ct).ConfigureAwait(false);
            return Ok(new ArtQueryOutDto { Data = "updated successfully" });
        }
        catch (Exception ex)
        {
            return StatusCode(500, new { detail = ex.Message });
        }
    }

    /// <summary>
    /// <c>GET /order/query</c>：逐单加载公司与明细、单价，计算 <c>TotalPrice</c>（教学示例；大数据量应改为单次 SQL 聚合）。
    /// </summary>
    [HttpGet("query")]
    public async Task<ActionResult<OrderQueryApiResponseDto>> Query(CancellationToken ct)
    {
        try
        {
            var orders = await db.Orders.AsNoTracking().OrderBy(o => o.Id).ToListAsync(ct).ConfigureAwait(false);
            var result = new List<OrderQueryRowOutDto>();

            foreach (var o in orders)
            {
                var companyName = await db.Companies.AsNoTracking()
                    .Where(c => c.Id == o.CompanyId)
                    .Select(c => c.Name)
                    .FirstOrDefaultAsync(ct)
                    .ConfigureAwait(false);

                var lines = await db.OrderLines.AsNoTracking().Where(l => l.OrderId == o.Id).ToListAsync(ct).ConfigureAwait(false);
                var plist = new List<OrderProductRowOutDto>();
                float total = 0;
                foreach (var line in lines)
                {
                    var prod = await db.Products.AsNoTracking().FirstOrDefaultAsync(p => p.Id == line.ProductId, ct).ConfigureAwait(false);
                    var price = prod?.Price ?? 0;
                    var num = line.Number ?? 0;
                    total += price * num;
                    plist.Add(new OrderProductRowOutDto
                    {
                        ProductId = line.ProductId,
                        ProductName = prod?.Name,
                        Number = line.Number,
                        Price = price
                    });
                }

                result.Add(new OrderQueryRowOutDto
                {
                    Id = o.Id,
                    OrderNumber = o.OrderNumber,
                    CompanyId = o.CompanyId,
                    CompanyName = companyName,
                    ProductList = plist,
                    TotalPrice = total
                });
            }

            return Ok(new OrderQueryApiResponseDto { Data = result });
        }
        catch (Exception ex)
        {
            return StatusCode(500, new { detail = ex.Message });
        }
    }

    /// <summary><c>DELETE /order/delete?id=</c>：先删 <c>order_product</c> 再删 <c>order</c>。</summary>
    [HttpDelete("delete")]
    public async Task<ActionResult<ArtQueryOutDto>> Delete([FromQuery] int id, CancellationToken ct)
    {
        try
        {
            await db.OrderLines.Where(l => l.OrderId == id).ExecuteDeleteAsync(ct).ConfigureAwait(false);
            await db.Orders.Where(o => o.Id == id).ExecuteDeleteAsync(ct).ConfigureAwait(false);
            return Ok(new ArtQueryOutDto { Data = "deleted successfully" });
        }
        catch (Exception ex)
        {
            return StatusCode(500, new { detail = ex.Message });
        }
    }
}
